exercises/practice/raindrops/.approaches/step-by-step/content.md

```diff
@@ -0,0 +1,77 @@
+# Step By Step
+
+```elixir
+defmodule Raindrops do
+  @spec convert(pos_integer) :: String.t()
+  def convert(number) do
+    pling = if rem(number, 3) == 0, do: "Pling", else: ""
+    plang = if rem(number, 5) == 0, do: "Plang", else: ""
+    plong = if rem(number, 7) == 0, do: "Plong", else: ""
+    sound = pling <> plang <> plong
+
+    if sound == "" do
+      Integer.to_string(number)
+    else
+      sound
+    end
+  end
+end
+```
+
+In this approach, we test each condition only once, similar to using the `case` on a tuple in the [pattern matching approach][pattern-matching-approach].
+However, this time, if a condition is true, we capture the sound component, and if it is not true, we capture the sound as an empty string.
+
+Once this is done, we can concatenate all three conditions to get the full sound.
+Finally, if the `sound` is empty, we can return the number or, alternatively, the calculated `sound`.
+
+## Functions
+
+We can create private functions to test for component sounds.
+
+```elixir
+defp pling(n) when rem(n, 3) == 0, do: "Pling"
+defp pling(_), do: ""
+defp plang(n) when rem(n, 5) == 0, do: "Plang"
+defp plang(_), do: ""
+defp plong(n) when rem(n, 7) == 0, do: "Plong"
+defp plong(_), do: ""
+defp sound(sound, number) when sound == "", do: Integer.to_string(number)
+defp sound(sound, _number), do: sound
+```
+
+Now the solution can look like this:
+```elixir
+def convert(number) do
+  sound(pling(number) <> plang(number) <> plong(number), number)
+end
+```
```

Comment:
String concatenation isn't used by Elixir devs that often. It has a big disadvantage - it crashes on `nil`s. Elixir devs usually always use string interpolation.

Using string interpolation in both of those solutions would make them slightly shorter:

```elixir
defmodule Raindrops do
  @spec convert(pos_integer) :: String.t()
  def convert(number) do
    pling = if rem(number, 3) == 0, do: "Pling"
    plang = if rem(number, 5) == 0, do: "Plang"
    plong = if rem(number, 7) == 0, do: "Plong"
    sound = "#{pling}#{plang}#{plong}"

    if sound == "" do
      Integer.to_string(number)
    else
      sound
    end
  end
end
```

```elixir
defp pling(n), do: if(rem(n, 3) == 0, do: "Pling")
defp plang(n), do: if(rem(n, 5) == 0, do: "Plang")
defp plong(n), do: if(rem(n, 7) == 0, do: "Plong")
defp sound(sound, number) when sound == "", do: Integer.to_string(number)
defp sound(sound, _number), do: sound

def convert(number) do
  sound("#{pling(number)}#{plang(number)}#{plong(number)}", number)
end
```